src/Interpreters/Context.cpp

```diff
@@ -1955,6 +1955,15 @@ void Context::killCurrentQuery() const
         elem->cancelQuery(true);
 }
 
+bool Context::isCurrentQueryKilled() const
+{
+    if (auto elem = getProcessListElement())
```

Comment:
afaik process list elem (weak ptr) expires after query finished and query context expired, but we sometimes destroy buffers after query context expired (I have seen that in stress test, happens when query had some exception). But in buffers desctructors we might sometimes finish the read (if prefetch was started but not waited for) so it will throw "Weak pointer to process_list_elem expired during query execution" from destructor which is the logical error. The error will be caught of course, but as it is a logical error it might sometimes be caught by our stress tests and cause terminate.

Reply:
I mean could there be a race between getting `auto query_context = CurrentThread::getQueryContext()` and this check, e.g. process list elem expired in between

Reply:
You are totally right here. Thanks for pointing out.

Process list element is created and destroyed during the request. Some buffers might have live time which has no synchronization with the corresponded process list element live time. 
However I claim that the buffers for writing parts to the discs have the enclosed lime time in corresponded process list element live time (buffer created and deleted during the presence of the process list element).
But `Context::isCurrentQueryKilled()` could be used theoretically in any place in any buffer, so it can't rely that process_list_elem is valid.